### DISCUSSION TITLE
Trigger job on tag push only

### DISCUSSION THREAD

**Discussion by msafi** [community/community#25692] [Actions]
<p>I have a workflow that is triggered by pushes, pull requests, and tags. The workflow has multiple jobs. One of the jobs is responsible for creating a release on GitHub. I want to restrict the release job to run only when the event is a created/pushed tag.</p>
<p>How can I do that?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/msafi">@msafi</a> ,</p>
<p>Creating/pushing a new tag to the repository is a type of <strong>Push event</strong>. By default, both pushing commits and pushing tags will trigger workflow run for, if you don’t set branches filter and tags filter for the Push event. More details you can reference <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" rel="nofollow noopener">here</a>.</p>
<p>When the wokflow run is triggered by pushing tag, the <strong>github.ref</strong> is the full name of the pushed tag ( <strong>refs/tags/&lt;tag&gt;</strong> ).<br>
So if you want to check if the workflow is triggered by pushing tag, you can try using the following <strong>if conditional</strong> for the job.</p>
<pre><code>if: github.event_name == 'push' &amp;&amp; contains(github.ref, 'refs/tags/')
</code></pre>
<p>Related docs:</p>
<p><a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github context</strong></a></p>
<p><a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#functions" rel="nofollow noopener"><strong>Functions</strong></a></p>

**Reply by JacobBrackett** [reply]
I'm seeing this commonly used wrong in different actions, it's important for anyone coming across this post to note that this is not a secure way to trigger an action that should only trigger when a new release is made, the [`release` event ](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) should be used. 

Anyone with write access to a repository can push a tag to any branch unless [tag protection](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-tag-protection-rules) is enabled, so just checking for the presence of a tag will allow users with write to bypass branch protection rules if the workflow is triggered on a push to any branch (e.g. `on: [push]`).

**Comment by kentquirk** [comment]
<p>I’ve spent the last two days on trying to make something like this work. Am I nuts or is this unreasonably hard?</p>
<p>Similar to you, all I want to be able to do is define some behavior that runs only when someone does a Release, and for that behavior to know the name of the branch it is running against.</p>
<p>I managed to make the first part of that work on something like the 30th attempt with an unholy and somewhat ridiculous combination of setting environment variables in one step using a shell, and then using them in a different step as new environment variables, and then finally using THOSE in an if: clause. As far as I can’t tell you can’t do that at the job level, but you can make individual steps conditional.</p>
<p>But once I got that working, it turned out that the branch name was no longer available because now the reference is to the release, not to the branch.</p>
<p>Sorry I don’t have better answers. This is my first experience with GitHub actions and I was really hoping that it would make it easy to convince my team to walk away from Jenkins, but my experience so far is that nothing has seemed obvious or easy. </p>
<p>One of the biggest problems is that all of the supposed solutions come with restrictions on their context. I’m sure those restrictions are there for engineering reasons but this doesn’t feel designed – it feels hacked. Dear GitHub team, this forum is full of people who are trying to figure out an easy way to access basic information about the context in which their jobs are running, and to use those in conditional behavior. Personally, I’m going to have to fall back to running python scripts and just invoking those with Actions.</p>

**Comment by msafi** [comment]
<p><a class="mention" href="/brightran">@brightran</a> Thanks for the solution!</p>
<p><a class="mention" href="/kentquirk">@kentquirk</a> Your experience sounds very similar to mine. I was banging my head against the wall for the past two days trying to figure things out. I found GitHub Actions to be very confusing and missing some basic things that I’ve come to expect from other CI services. I’m hoping GitHub Actions will continue to mature and become easier with time.</p>

**Comment by dbrgn** [comment]
<p>So if I want to run some jobs on push to master, on push to a v* tag or on a pull request, I have to use the following if condition?</p>
<pre><code>on:
    push:
    pull_request:

jobs:
    build:
        if: github.event_name == 'pull_request' || (github.event_name == 'push' &amp;&amp; (contains(github.ref, '/heads/master') || contains(github.ref, '/tags/v')))
        ...
</code></pre>
<p>Seems pretty awkward. Maybe there’s a better solution?</p>

**Comment by rachelcarmena** [comment]
<p>Thanks for the tip <a class="mention" href="/brightran">@brightran</a> ! However, it seems the push event is just triggered when pushing a tag for one of the last 3 commits. When creating a tag for an older commit and pushing it (just pushing a single tag), the push event is not triggered . Please, do you know how to trigger the push event for older commits?</p>

**Comment by reuben** [comment]
<p>That would likely trigger a <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#create" rel="noopener nofollow ugc">create</a> event instead (but I haven’t tested it).</p>

**Comment by justinmchase** [comment]
You can filter on tags specifically now

```yml
on:
  push:
    tags:
      - '*'
```

**Reply by le91688** [reply]
Im unsure this helps. Unless im mistaken, `on:` pertains to triggering the _workflow_. What OP wants is for jobs within a workflow to only happen if tag/release . 

For example, on every push, lint the code, and if that push happens to be a tag / release, publish to Pypi.  You would not want Linting to run only on tags .

**Reply by justinmchase** [reply]
Just make two workflows, one that handles the above events and another that handles the other events. If they do the same thing then make a 3rd workflow with the shared code and use workflow_call with inputs to invoke the shared workflow and send discrete signals.

**Reply by TheButlah** [reply]
I thought you couldn't share artifacts between workflows

**Reply by justinmchase** [reply]
Workflows and jobs have inputs and outputs and you can also store artifacts and fetch them again in another job. 

https://github.com/actions/upload-artifact

**Reply by DEVigodE** [reply]
Perfecto
